qualtran/_infra/gate_with_registers.py

```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
 
         wire_symbols[0] = self.__class__.__name__
         return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
+
+
+def get_ctrl_system_for_single_qubit_controlled(
+    bloq: GateWithRegisters, ctrl_spec: Optional['CtrlSpec'] = None
+) -> Tuple['Bloq', 'AddControlledT']:
+    """Wrapper to correctly wire a custom implementation of a single-qubit controlled version of a gate.
+
+    For any gate that supports a single-qubit controlled version
+    with an API that has `control_val` and `control_registers`.
+    Add the `get_ctrl_system` implementation as shown below:
+
+    .. code:: python
+
+        class MyGate(GateWithRegisters):
+            control_val: Optional[int]
+
+            @property
+            def control_registers() -> Tuple[Register, ...]:
+                ...
+
+            def get_ctrl_system(
+                self, ctrl_spec: Optional['CtrlSpec'] = None
+            ) -> Tuple['Bloq', 'AddControlledT']:
+                from qualtran._infra.gate_with_registers import get_ctrl_system_for_single_qubit_controlled
+
+                return get_ctrl_system_for_single_qubit_controlled(self, ctrl_spec)
+    """
+    if ctrl_spec is None:
+        ctrl_spec = CtrlSpec()
+
+    assert hasattr(bloq, 'control_val')
+    assert hasattr(bloq, 'control_registers')
```

Comment:
I don't know if this is a better or worse idea, but: did you consider making this a mixin?

Reply:
this does make the code cleaner, and enables type-checking. I made a class `SpecializedSingleQubitControlledGate` (similar to the design of `UnaryIterationGate`), please have a look